MongoSourcesWindowedTest.testWindows [HZ-2558]

PR builder
https://github.com/hazelcast/hazelcast/pull/24321#pullrequestreview-1397524153
https://jenkins.hazelcast.com/job/Hazelcast-pr-builder/18029/testReport/junit/com.hazelcast.jet.mongodb/MongoSourcesWindowedTest/testWindows/

<details><summary>Stacktrace:</summary>

```
org.junit.ComparisonFailure: [Sum of windows is equal to input element count] expected:<[100]L> but was:<[99]L>
	at sun.reflect.GeneratedConstructorAccessor29.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at com.hazelcast.jet.mongodb.MongoSourcesWindowedTest.lambda$testWindows$2(MongoSourcesWindowedTest.java:93)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1288)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1392)
	at com.hazelcast.jet.mongodb.MongoSourcesWindowedTest.testWindows(MongoSourcesWindowedTest.java:84)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:750)
```

</details>

<details><summary>Standard output:</summary>

```
09:33:58,495  INFO || - [PartitionStateManager] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Initializing cluster partition table arrangement...
09:33:58,509 DEBUG || - [JobClassLoaderService] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Creating job classLoader for job 09c0-e614-55c0-0001
09:33:58,509 DEBUG || - [JobClassLoaderService] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Create processor classloader map for job 09c0-e614-55c0-0001
09:33:58,529 DEBUG || - [JobCoordinationService] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-2 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partition replication is not in safe state, but in REPLICA_NOT_SYNC
09:33:58,530 DEBUG || - [Planner] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-1 - Watermarks in the pipeline will be throttled to 5
09:33:58,554  INFO || - [JobCoordinationService] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Starting job 09c0-e614-55c0-0001 based on submit request
09:33:58,565  INFO || - [MasterJobContext] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Didn't find any snapshot to restore for job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001
09:33:58,565  INFO || - [MasterJobContext] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Start executing job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001, execution graph in DOT format:
digraph DAG {
	"MongoSource(testDefaultDatabase133/testWindows)" [localParallelism=2];
	"sliding-window-prepare" [localParallelism=16];
	"sliding-window" [localParallelism=16];
	"listSink(result)" [localParallelism=1];
	"MongoSource(testDefaultDatabase133/testWindows)" -> "sliding-window-prepare" [label="partitioned", queueSize=1024];
	subgraph cluster_0 {
		"sliding-window-prepare" -> "sliding-window" [label="distributed-partitioned", queueSize=1024];
	}
	"sliding-window" -> "listSink(result)" [queueSize=1024];
}
HINT: You can use graphviz or http://viz-js.com to visualize the printed graph.
09:33:58,565 DEBUG || - [MasterJobContext] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Building execution plan for job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001
09:33:58,565 DEBUG || - [MasterJobContext] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Built execution plans for job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001
09:33:58,570 DEBUG || - [InitExecutionOperation] hz.MongoSourcesWindowedTest_heuristic_mendeleev.generic-operation.thread-4 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Initializing execution plan for job 09c0-e614-55c0-0001, execution 09c0-e614-55c1-0001 from [127.0.0.1]:5701
09:33:58,570 DEBUG || - [InitExecutionOperation] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Initializing execution plan for job 09c0-e614-55c0-0001, execution 09c0-e614-55c1-0001 from [127.0.0.1]:5701
09:33:58,570 DEBUG || - [JobClassLoaderService] hz.MongoSourcesWindowedTest_heuristic_mendeleev.generic-operation.thread-4 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Creating job classLoader for job 09c0-e614-55c0-0001
09:33:58,570 DEBUG || - [JobClassLoaderService] hz.MongoSourcesWindowedTest_heuristic_mendeleev.generic-operation.thread-4 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Create processor classloader map for job 09c0-e614-55c0-0001
09:33:58,596  INFO || - [JobExecutionService] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-2 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Execution plan for jobId=09c0-e614-55c0-0001, jobName='09c0-e614-55c0-0001', executionId=09c0-e614-55c1-0001 initialized
09:33:58,596  INFO || - [JobExecutionService] hz.MongoSourcesWindowedTest_heuristic_mendeleev.cached.thread-3 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Execution plan for jobId=09c0-e614-55c0-0001, jobName='09c0-e614-55c0-0001', executionId=09c0-e614-55c1-0001 initialized
09:33:58,596 DEBUG || - [MasterJobContext] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Init of job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001 was successful
09:33:58,596 DEBUG || - [MasterJobContext] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Executing job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001
09:33:58,597  INFO || - [JobExecutionService] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Start execution of job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001 from coordinator [127.0.0.1]:5701
09:33:58,597  INFO || - [JobExecutionService] hz.MongoSourcesWindowedTest_heuristic_mendeleev.generic-operation.thread-5 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Start execution of job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001 from coordinator [127.0.0.1]:5701
09:33:58,603 DEBUG || - [ReadMongoP] hz.MongoSourcesWindowedTest_brave_mendeleev.jet.blocking.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] [09c0-e614-55c0-0001/MongoSource(testDefaultDatabase133/testWindows)#3] (Re)connecting to MongoDB
09:33:58,603 DEBUG || - [ReadMongoP] hz.MongoSourcesWindowedTest_heuristic_mendeleev.jet.blocking.thread-0 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] [09c0-e614-55c0-0001/MongoSource(testDefaultDatabase133/testWindows)#0] (Re)connecting to MongoDB
09:33:58,604 DEBUG || - [ReadMongoP] hz.MongoSourcesWindowedTest_heuristic_mendeleev.jet.blocking.thread-1 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] [09c0-e614-55c0-0001/MongoSource(testDefaultDatabase133/testWindows)#1] (Re)connecting to MongoDB
09:33:58,607 DEBUG || - [ReadMongoP] hz.MongoSourcesWindowedTest_brave_mendeleev.jet.blocking.thread-0 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] [09c0-e614-55c0-0001/MongoSource(testDefaultDatabase133/testWindows)#2] (Re)connecting to MongoDB
09:33:58,629 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 3ms
09:34:03,625 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-8 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 2ms
09:34:08,345 DEBUG || - [MasterJobContext] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-11 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not scaling up job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001: not running or already running on all members
09:34:08,628 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-11 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:34:13,629 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-18 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:34:18,627 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-10 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 2ms
09:34:23,623 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-10 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:34:28,627 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-10 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 3ms
09:34:33,625 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-10 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 2ms
09:34:38,625 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-10 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 2ms
09:34:43,624 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-17 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:34:48,623 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-17 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:34:53,623 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-11 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:34:58,622 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-17 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:35:03,624 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-8 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:35:08,625 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-15 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 2ms
09:35:13,622 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-8 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:35:18,624 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-14 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:35:23,626 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-17 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 2ms
09:35:28,623 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-10 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 2ms
09:35:33,622 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-12 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:35:38,623 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 2ms
09:35:43,623 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:35:48,622 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:35:53,622 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-15 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:35:59,567 DEBUG || - [JobRepository] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-8 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 1ms
09:35:59,651  INFO |testWindows| - [SimpleTestInClusterSupport] Thread-6 - Ditching 1 jobs in SimpleTestInClusterSupport.@After: [09c0-e614-55c0-0001]
09:35:59,651 DEBUG |testWindows| - [AbstractJobProxy] Thread-6 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Sending CANCEL_FORCEFUL request for job 09c0-e614-55c0-0001 (name ??)
09:35:59,690 DEBUG || - [JobExecutionService] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-14 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Completed execution of job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001
09:35:59,690 DEBUG || - [JobClassLoaderService] hz.MongoSourcesWindowedTest_heuristic_mendeleev.cached.thread-13 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Finish JobClassLoaders phaseCount = 0, removing classloaders for jobId=09c0-e614-55c0-0001
09:35:59,690 DEBUG || - [JobExecutionService] hz.MongoSourcesWindowedTest_heuristic_mendeleev.cached.thread-13 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Completed execution of job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001
09:35:59,690 DEBUG || - [JobExecutionService] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-14 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Execution of job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001 completed with failure
java.util.concurrent.CompletionException: com.hazelcast.jet.impl.exception.JobTerminateRequestedException: CANCEL_FORCEFUL
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:783) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990) ~[?:1.8.0_351]
	at com.hazelcast.jet.impl.util.NonCompletableFuture.internalCompleteExceptionally(NonCompletableFuture.java:72) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$ExecutionTracker.taskletDone(TaskletExecutionService.java:498) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:330) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_351]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_351]
	at java.lang.Thread.run(Thread.java:750) ~[?:1.8.0_351]
Caused by: com.hazelcast.jet.impl.exception.JobTerminateRequestedException: CANCEL_FORCEFUL
	at com.hazelcast.jet.impl.execution.ExecutionContext.terminateExecution(ExecutionContext.java:318) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.JobExecutionService.terminateExecution0(JobExecutionService.java:723) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.JobExecutionService.terminateExecution(JobExecutionService.java:719) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.operation.TerminateExecutionOperation.doRun(TerminateExecutionOperation.java:59) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.operation.AsyncOperation.run(AsyncOperation.java:55) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.Operation.call(Operation.java:192) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:299) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:270) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:219) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.run(OperationExecutorImpl.java:454) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.runOrExecute(OperationExecutorImpl.java:511) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvokeLocal(Invocation.java:618) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:597) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:558) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:241) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.impl.InvocationBuilderImpl.invoke(InvocationBuilderImpl.java:71) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.MasterContext.invokeOnParticipant(MasterContext.java:343) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.MasterContext.invokeOnParticipants(MasterContext.java:326) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.MasterJobContext.lambda$cancelExecutionInvocations$25(MasterJobContext.java:725) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_351]
	at java.lang.Thread.run(Thread.java:750) [?:1.8.0_351]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76) [hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:111) [hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
09:35:59,690 DEBUG || - [JobExecutionService] hz.MongoSourcesWindowedTest_heuristic_mendeleev.cached.thread-13 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Execution of job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001 completed with failure
java.util.concurrent.CompletionException: com.hazelcast.jet.impl.exception.JobTerminateRequestedException: CANCEL_FORCEFUL
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:783) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990) ~[?:1.8.0_351]
	at com.hazelcast.jet.impl.util.NonCompletableFuture.internalCompleteExceptionally(NonCompletableFuture.java:72) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$ExecutionTracker.taskletDone(TaskletExecutionService.java:498) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:330) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_351]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_351]
	at java.lang.Thread.run(Thread.java:750) ~[?:1.8.0_351]
Caused by: com.hazelcast.jet.impl.exception.JobTerminateRequestedException: CANCEL_FORCEFUL
	at com.hazelcast.jet.impl.execution.ExecutionContext.terminateExecution(ExecutionContext.java:318) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.JobExecutionService.terminateExecution0(JobExecutionService.java:723) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.JobExecutionService.terminateExecution(JobExecutionService.java:719) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.operation.TerminateExecutionOperation.doRun(TerminateExecutionOperation.java:59) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.operation.AsyncOperation.run(AsyncOperation.java:55) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.Operation.call(Operation.java:192) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:299) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:270) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:502) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:202) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:142) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.loop(OperationThread.java:134) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.executeRun(OperationThread.java:115) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:111) [hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
09:35:59,691 DEBUG || - [MasterJobContext] ForkJoinPool.commonPool-worker-2 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001 received response to StartExecutionOperation from [127.0.0.1]:5701: com.hazelcast.jet.impl.exception.JobTerminateRequestedException: CANCEL_FORCEFUL
09:35:59,692 DEBUG || - [MasterJobContext] ForkJoinPool.commonPool-worker-2 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001 received response to StartExecutionOperation from [127.0.0.1]:5702: com.hazelcast.jet.impl.exception.JobTerminateRequestedException: CANCEL_FORCEFUL
09:35:59,692 DEBUG || - [MasterJobContext] ForkJoinPool.commonPool-worker-2 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001 to be cancelled after Execution
09:35:59,692 DEBUG || - [JobClassLoaderService] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-10 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Finish JobClassLoaders phaseCount = 0, removing classloaders for jobId=09c0-e614-55c0-0001
09:35:59,692  INFO || - [MasterJobContext] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-10 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Execution of job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001 got terminated, reason=com.hazelcast.jet.impl.exception.CancellationByUserException
	Start time: 2023-04-24T09:33:58.554
	Duration: 00:02:01.138
	To see additional job metrics enable JobConfig.storeMetricsAfterJobCompletion
09:35:59,694 DEBUG || - [JobCoordinationService] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-10 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] job '09c0-e614-55c0-0001', execution 09c0-e614-55c1-0001 is completed
09:35:59,694  WARN || - [JoinSubmittedJobOperation] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-10 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] null
com.hazelcast.jet.impl.exception.CancellationByUserException: null
	at com.hazelcast.jet.impl.MasterJobContext.createCancellationException(MasterJobContext.java:211) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.MasterJobContext.getErrorFromResponses(MasterJobContext.java:626) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.MasterJobContext.lambda$invokeStartExecution$12(MasterJobContext.java:576) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.MasterContext.lambda$invokeOnParticipant$3(MasterContext.java:361) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.util.ExceptionUtil.lambda$withTryCatch$0(ExceptionUtil.java:180) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.AbstractInvocationFuture$WhenCompleteNode.lambda$execute$0(AbstractInvocationFuture.java:1570) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.test.TestLoggingUtils$CustomTestNameAwareForkJoinPool$TestNameAwareRunnable.run(TestLoggingUtils.java:104) ~[hazelcast-5.3.0-SNAPSHOT-tests.jar:5.3.0-SNAPSHOT]
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402) ~[?:1.8.0_351]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) ~[?:1.8.0_351]
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1067) ~[?:1.8.0_351]
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1703) ~[?:1.8.0_351]
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:172) ~[?:1.8.0_351]
09:35:59,695  WARN || - [JoinSubmittedJobOperation] hz.MongoSourcesWindowedTest_brave_mendeleev.cached.thread-10 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] null
com.hazelcast.jet.impl.exception.CancellationByUserException: null
	at com.hazelcast.jet.impl.MasterJobContext.createCancellationException(MasterJobContext.java:211) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.MasterJobContext.getErrorFromResponses(MasterJobContext.java:626) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.MasterJobContext.lambda$invokeStartExecution$12(MasterJobContext.java:576) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.MasterContext.lambda$invokeOnParticipant$3(MasterContext.java:361) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.jet.impl.util.ExceptionUtil.lambda$withTryCatch$0(ExceptionUtil.java:180) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.spi.impl.AbstractInvocationFuture$WhenCompleteNode.lambda$execute$0(AbstractInvocationFuture.java:1570) ~[hazelcast-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at com.hazelcast.test.TestLoggingUtils$CustomTestNameAwareForkJoinPool$TestNameAwareRunnable.run(TestLoggingUtils.java:104) ~[hazelcast-5.3.0-SNAPSHOT-tests.jar:5.3.0-SNAPSHOT]
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402) ~[?:1.8.0_351]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) ~[?:1.8.0_351]
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1067) ~[?:1.8.0_351]
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1703) ~[?:1.8.0_351]
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:172) ~[?:1.8.0_351]
09:35:59,697  INFO |testWindows| - [SimpleTestInClusterSupport] Thread-6 - Destroying 5 distributed objects in SimpleTestInClusterSupport.@After: [hz:impl:listService/result, hz:impl:mapService/__jet.executionRecords, hz:impl:mapService/__jet.results, hz:impl:mapService/__jet.records, hz:impl:flakeIdGeneratorService/__jet.ids]
BuildInfo right after testWindows(com.hazelcast.jet.mongodb.MongoSourcesWindowedTest): BuildInfo{version='5.3.0-SNAPSHOT', build='20230424', buildNumber=20230424, revision=2cc95e6, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'testWindows(com.hazelcast.jet.mongodb.MongoSourcesWindowedTest):'
09:33:55, accumulated pauses: 31 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:34:00, accumulated pauses: 26 ms, max pause: 2 ms, pauses over 1000 ms: 0
09:34:05, accumulated pauses: 24 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:34:10, accumulated pauses: 26 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:34:15, accumulated pauses: 30 ms, max pause: 1 ms, pauses over 1000 ms: 0
09:34:20, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:34:25, accumulated pauses: 32 ms, max pause: 1 ms, pauses over 1000 ms: 0
09:34:30, accumulated pauses: 32 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:34:35, accumulated pauses: 32 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:34:40, accumulated pauses: 32 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:34:45, accumulated pauses: 63 ms, max pause: 31 ms, pauses over 1000 ms: 0
09:34:50, accumulated pauses: 30 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:34:55, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:35:00, accumulated pauses: 33 ms, max pause: 2 ms, pauses over 1000 ms: 0
09:35:05, accumulated pauses: 30 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:35:10, accumulated pauses: 31 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:35:15, accumulated pauses: 31 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:35:20, accumulated pauses: 31 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:35:25, accumulated pauses: 30 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:35:30, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:35:35, accumulated pauses: 30 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:35:40, accumulated pauses: 32 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:35:45, accumulated pauses: 31 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:35:50, accumulated pauses: 32 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:35:55, accumulated pauses: 968 ms, max pause: 946 ms, pauses over 1000 ms: 0


No metrics recorded during the test
```

</details>

--- a/extensions/mongodb/src/test/java/com/hazelcast/jet/mongodb/MongoSourcesWindowedTest.java
+++ b/extensions/mongodb/src/test/java/com/hazelcast/jet/mongodb/MongoSourcesWindowedTest.java
@@ -17,6 +17,7 @@
 package com.hazelcast.jet.mongodb;
 
 import com.hazelcast.collection.IList;
+import com.hazelcast.jet.Job;
 import com.hazelcast.jet.datamodel.KeyedWindowResult;
 import com.hazelcast.jet.datamodel.WindowResult;
 import com.hazelcast.jet.pipeline.Pipeline;
@@ -34,6 +35,7 @@ import org.junit.runner.RunWith;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import static com.hazelcast.jet.aggregate.AggregateOperations.counting;
+import static com.hazelcast.jet.core.JobStatus.RUNNING;
 import static com.hazelcast.jet.pipeline.Sinks.list;
 import static com.hazelcast.jet.pipeline.WindowDefinition.tumbling;
 import static com.hazelcast.test.DockerTestUtil.assumeDockerEnabled;
@@ -60,16 +62,18 @@ public class MongoSourcesWindowedTest extends AbstractMongoTest {
         Pipeline pipeline = Pipeline.create();
         pipeline.readFrom(streamSource)
                 .withNativeTimestamps(0).setLocalParallelism(2)
+                .peek()
                 .groupingKey(d -> d.getString("test"))
                 .window(tumbling(5))
                 .aggregate(counting())
+                .peek()
                 .writeTo(list(result));
 
         MongoCollection<Document> collection =
                 mongo.getDatabase(defaultDatabase()).getCollection(testName.getMethodName());
         AtomicInteger counter = new AtomicInteger(0);
         spawn(() -> {
-            while (counter.get() < 100) {
+            while (counter.get() < 20) {
                 ObjectId key = ObjectId.get();
                 collection.insertOne(new Document("test", "testowe").append("_id", key));
                 counter.incrementAndGet();
@@ -79,7 +83,8 @@ public class MongoSourcesWindowedTest extends AbstractMongoTest {
                 collection.insertOne(new Document("test", "other").append("_id", key));
             }
         });
-        instance().getJet().newJob(pipeline);
+        Job job = instance().getJet().newJob(pipeline);
+        assertJobStatusEventually(job, RUNNING);
 
         assertTrueEventually(() -> {
             assertThat(result).isNotEmpty();
@@ -91,6 +96,6 @@ public class MongoSourcesWindowedTest extends AbstractMongoTest {
             assertThat(currentSum)
                     .as("Sum of windows is equal to input element count")
                     .isEqualTo(counter.get());
-        });
+        }, 20);
     }
 }

